### PR TITLE
Send rallypoints and geofences through mavros

### PR DESCRIPTION
**Pull request by Charlie-Burge**
For #1157. Breaks out the mission protocol into a base class from the waypoint plugin and adds the ability to send rallypoints and geofences

### CODE REVIEW

**Review by vooon — CHANGES_REQUESTED** [review]
Please keep past authors copyright. But code looks good to me.
did you test it already?

File: mavros/include/mavros/mission_protocol_base.h [hunk 1]
```diff
@@ -0,0 +1,599 @@
+/**
+ * @brief Mission base plugin
+ * @file mission_protocol_base.h
+ * @author Charlie Burge <charlieburge@yahoo.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2021 Charlie Burge.
```
[the remark this answers is not in the document]
Comment (Charlie-Burge) [reply mavros/include/mavros/mission_protocol_base.h:12]: Of course - changes made :)

File: mavros/include/mavros/mission_protocol_base.h [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+	//global_names = [v for v in all_names if v.startswith('GLOBAL')]
+	//local_names = [v for v in all_names if v.startswith(('LOCAL', 'BODY', 'MOCAP', 'VISION', 'ESTIM'))]
+	//other_names = ['MISSION']
+	//cog.outl("switch(frame){")
+	//for names, factor in [(global_names, 10000000), (local_names, 10000), (other_names, 1)]:
+	//	for name in names:
+	//		cog.outl(f"case enum_value(MAV_FRAME::{name}):")
+	//	cog.outl(f"\treturn {factor};")
+	//cog.outl("default:\n\treturn 1;")
+	// ]]]
```
[the remark this answers is not in the document]
Comment (Charlie-Burge) [reply mavros/include/mavros/mission_protocol_base.h:61]: Great thanks! I had just copied from waypoint.cpp, but hadn't regenerated since

**Comment by Charlie-Burge** [comment]
> did you test it already?

I have tested using ArduPilot, but haven't with PX4. I tested with ArduCopter SITL using mavproxy, and using a CubeOrange and CubeBlack running the latest stable release of ArduCopter. My testing has not been extensive, although I have tried all the functionality, so please do check my work. 

From testing I noticed when pulling the rallypoints or geofence via the service it would trigger a scheduled pull of the mission waypoints. This only occurred when using maxproxy and not when testing using the CubeOrange and CubeBlack connected directly to mavros via USB. I believe this is a subtlety in the fact both mavproxy and mavros believe each other is the GCS - so schedule a pull after each other.

**Review by vooon — APPROVED** [review]
LGTM, Thanks!

**Pull request merged**